File: docs/chapters/04-data-structures-and-complexity/java/LruCacheSession.java
// Scenario: cap in-memory checkout sessions per node (10k) with LRU eviction
// Demonstrates: O(1) get/put via LinkedHashMap access-order
// Trade-off: not thread-safe; production shards by sessionId or uses Caffeine

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

final class LruCacheSession<V> {
    private final int capacity;
    private final LinkedHashMap<String, V> map;

    LruCacheSession(int capacity) {
        this.capacity = capacity;
        this.map = new LinkedHashMap<>(capacity, 0.75f, true) {
            @Override protected boolean removeEldestEntry(Map.Entry<String, V> eldest) {
                return size() > LruCacheSession.this.capacity;
            }
        };
    }

    Optional<V> get(String sessionId) {
        return Optional.ofNullable(map.get(sessionId));
    }

    void put(String sessionId, V state) {
        map.put(sessionId, state);
    }

    int size() { return map.size(); }
}

File: docs/chapters/04-data-structures-and-complexity/go/lru_cache_session.go
// Scenario: cap in-memory checkout sessions per node (10k) with LRU eviction
// Demonstrates: O(1) get/put via map + container/list order
// Trade-off: explicit list vs sync.Map; shard or use ristretto/Caffeine at scale

package main

import "container/list"

type LruCacheSession[V any] struct {
	cap   int
	ll    *list.List
	items map[string]*list.Element
}

type entry[V any] struct {
	key string
	val V
}

func NewLruCacheSession[V any](capacity int) *LruCacheSession[V] {
	return &LruCacheSession[V]{
		cap:   capacity,
		ll:    list.New(),
		items: make(map[string]*list.Element, capacity),
	}
}

func (c *LruCacheSession[V]) Get(key string) (V, bool) {
	if el, ok := c.items[key]; ok {
		c.ll.MoveToFront(el)
		return el.Value.(*entry[V]).val, true
	}
	var zero V
	return zero, false
}

func (c *LruCacheSession[V]) Put(key string, val V) {
	if el, ok := c.items[key]; ok {
		c.ll.MoveToFront(el)
		el.Value.(*entry[V]).val = val
		return
	}
	el := c.ll.PushFront(&entry[V]{key: key, val: val})
	c.items[key] = el
	if c.ll.Len() > c.cap {
		back := c.ll.Back()
		c.ll.Remove(back)
		delete(c.items, back.Value.(*entry[V]).key)
	}
}
